utilities/env_librados.cc

```diff
@@ -0,0 +1,1043 @@
+}
+
+rocksdb::Status EnvLibrados::_GetSubFnamesByOmap(
+    const std::string& dir,
+    std::vector<std::string> * result
+    ) {
+    std::string start_after(dir);
+    std::string filter_prefix(dir);
+    std::map<std::string, librados::bufferlist> kvs;
+    _io_ctx.omap_get_vals(_db_id, 
+        start_after, filter_prefix, 
+        MAX_ITEMS_IN_FS, &kvs);
+
+    result->clear();
+    for (auto i = kvs.begin(); i != kvs.end(); i++) {
+        LOG_DEBUG("[match]%s\n", i->first.c_str());
+        if(i->first.size() > dir.size() 
+                && i->first[dir.size()] == '/'
+                && std::string::npos == i->first.find('/', dir.size()+1)) {
+            result->push_back(i->first.substr(dir.size()+1));
+        }
+    }
+    return rocksdb::Status::OK();
+}
+
+rocksdb::Status EnvLibrados::_DelFidByOmap(
+    const std::string& fname) {
+    LOG_DEBUG("%s\n", fname.c_str());
+    std::set<std::string> keys;
+    keys.insert(fname);
+    int r = _io_ctx.omap_rm_keys(_db_id, keys);
+    return err_to_status(r);
+}
+
+
+
+rocksdb::Status EnvLibrados::_GetFid(
+        const std::string &fname, 
+        std::string& fid) {
+    return _GetFidByOmap(fname, fid);
+}
+rocksdb::Status EnvLibrados::_AddFid(
+        const std::string& fname, 
+        const std::string& fid) {
+    return _AddFidByOmap(fname, fid);
+}
+
+rocksdb::Status EnvLibrados::_GetSubFnames(
+    const std::string& dir,
+    std::vector<std::string> * result
+    ) {
+    return _GetSubFnamesByOmap(dir, result);
+}
+
+rocksdb::Status EnvLibrados::_DelFid(
+    const std::string& fname) {
+    return _DelFidByOmap(fname);
+}
+
+
+
+std::string EnvLibrados::GenerateUniqueId() {
+/*
+  std::string uuid_file = "/proc/sys/kernel/random/uuid";
+
+  int ret = access(uuid_file.c_str(), F_OK);
+  if (0 == ret) {
+    std::string uuid;
+    rocksdb::Status s = ReadFileToString(this, uuid_file, &uuid);
+    if (s.ok()) {
+      return uuid;
+    }
+  }
+*/  
+  // Could not read uuid_file - generate uuid using "nanos-random"
+  rocksdb::Random64 r(time(nullptr));
+  uint64_t random_uuid_portion =
+    r.Uniform(std::numeric_limits<uint64_t>::max());
+  uint64_t nanos_uuid_portion = NowNanos();
+  char uuid2[200];
+  snprintf(uuid2,
+           200,
+           "%lx-%lx",
+           (unsigned long)nanos_uuid_portion,
+           (unsigned long)random_uuid_portion);
+  return uuid2;
+}
+
+
+rocksdb::Status EnvLibrados::NewSequentialFile(
+  const std::string& fname,
+  std::unique_ptr<rocksdb::SequentialFile>* result,
+  const rocksdb::EnvOptions& options)
+{
+    std::string dir, file, fid;
+    split(fname, &dir, &file);
+    rocksdb::Status s = _GetFid(dir, fid);
+
+    if (!s.ok() || fid != DIR_TAG_VALUE) {
+        return rocksdb::Status::IOError();
+    }
+
+    s = _GetFid(fname, fid);
+
+    if (rocksdb::Status::NotFound() == s) {
+        return rocksdb::Status::IOError();
+    }
+
+    result->reset(new LibradosSequentialFile(_io_ctx, fid));
```

Comment:
I choose to pass the pointer of io_ctx to the file handle objects. It seems fair enough right now.